src/main/java/com/atlassian/db/replica/spi/ClusterConsistency.java

```diff
@@ -0,0 +1,36 @@
+package com.atlassian.db.replica.spi;
+
+import com.atlassian.db.replica.api.Database;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Collection;
+
+public interface ClusterConsistency {
```

Comment:
Something is wrong with the API. Even if we use a cluster, it's enough to use ReplicaConsistency for CBR (check before read). `ClusterConsistency` seems to be a good fit only for cluster + WAW (Synchronous writes). It seems it's not a good abstraction to replace ReplicaConsistency. Maybe it's worth keeping ReplicaConsistency as a consistency API and see how it evolves in the future.

Reply:
@wyrzyk Why do we need ClusterConsistency? If you combine replica discoverer and ReplicaConsistency you have the ability to check consistency for each replica / filter only consistent replicas etc. Do we really need contracts for cluster?

Reply:
>  If you combine replica discoverer and ReplicaConsistency 

I don't want to introduce discoverer API. I'm still working on the PR to simplify it more. I think the new API will be only a new method in ConnectionProvider (generalized getReplicaConnection to getReplicas) + ClusterConsistency. ClusterConsistency may not be a good name 🤔. I like the change of return type from boolean to Optional<Database>. It seems to enable more use-cases:
- CBR can either use AWS LB to resolve the connection or use its own LB and provide one of the already consistent replicas.
- Same for WAW, we don't need to wait for all replicas. We just need to ensure ConsistencyProvider write and getConsistent are consistent (if it waits only for 3 replica nodes, it can LB only between the 3 most up to date replicas).

Reply:
> if it waits only for 3 replica nodes, it can LB only between the 3 most up to date replicas
Does it make sense when it comes to distributed systems? I think about situations where you have multiple-nodes. Local guarding is not enough if you want to keep consistency.

Reply:
> Local guarding is not enough if you want to keep consistency

Is it local when with replica discovery? All nodes see the same state of cluster topology.

Reply:
> We just need to ensure ConsistencyProvider write and getConsistent are consistent
Hmm, this would make WAW to check before reading. Which makes it CBR, right?

Reply:
It would be WAW but wait for some replicas instead of wait for all.

Reply:
Anyway, the API is still broken and will evolve.

Reply:
> It would be WAW but wait for some replicas instead of wait for all.
How would we guarantee that we only read from consistent replicas? It would require some checking before reading, right?